src/sentry/grouping/parameterization.py

```diff
@@ -0,0 +1,368 @@
+import dataclasses
+import re
+from collections import defaultdict
+from collections.abc import Callable, Sequence
+from functools import lru_cache
+
+import tiktoken
+
+__all__ = [
+    "ParameterizationCallable",
+    "ParameterizationCallableExperiment",
+    "ParameterizationExperiment",
+    "ParameterizationRegex",
+    "ParameterizationRegexExperiment",
+    "Parameterizer",
+    "UniqueIdExperiment",
+]
+
+
+@dataclasses.dataclass
+class ParameterizationRegex:
+
+    name: str  # name of the pattern also used as group name in combined regex
+    raw_pattern: str  # regex pattern w/o matching group name
+    lookbehind: str | None = None  # positive lookbehind prefix if needed
+    lookahead: str | None = None  # positive lookahead postfix if needed
+    counter: int = 0
+
+    # These need to be used with `(?x)` tells the regex compiler to ignore comments
+    # and unescaped whitespace, so we can use newlines and indentation for better legibility.
+
+    @property
+    def pattern(self) -> str:
+        """
+        Returns the regex pattern for with as a named matching group and lookbehind/lookahead if needed.
+        """
+        prefix = rf"(?<={self.lookbehind})" if self.lookbehind else ""
+        postfix = rf"(?={self.lookahead})" if self.lookahead else ""
+        return rf"{prefix}(?P<{self.name}>{self.raw_pattern}){postfix}"
+
+    @property
+    def compiled_pattern(self) -> re.Pattern[str]:
+        """
+        Returns the compiled regex pattern for with as a named matching group and lookbehind/lookahead if needed.
+        """
+        if not hasattr(self, "_compiled_pattern"):
+            self._compiled_pattern = re.compile(rf"(?x){self.pattern}")
```

Comment:
I can't remember what yellow meant in coverage but just FYI
<img width="532" alt="image" src="https://github.com/getsentry/sentry/assets/44410/7ca9a40e-9a8c-4c89-a81e-c83124992fd8">

Reply:
it's yellow, because the only path where it's not cached is covered